<a href="https://colab.research.google.com/github/matthewreed2000/cse380-notebooks/blob/master/ponder_and_prove_combinatorics_and_probability.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Ponder and Prove Combinatorics and Probability
#### Due: Saturday, 6 February 2021, 11:59 pm.

*Matthew Reed*

In collaboration with:
- Davis Kerr
- Paul Semadeni
- Brayden Whitlock

## Conjecture

A number-theoretic conjecture of combinatorial significance is the following:

$degree2({2n \choose n}) =$ the "bits-on count" (or population count, or Hamming weight) of $n$.

$degree2(m)$ is defined as the number (degree, exponent) of 2's in the prime factorization of $m$.

In other words, for any $m$, a positive integer, $m = 2^e \cdot o$ where $o$ is an odd positive integer (could be 1) and $e$ is a natural number, including zero --- which would be the case when $m$ is odd. It's the $e$ that is the $degree2$ of $m$.

Another way to state this conjecture is that the number of 1's in the binary expansion of ${2n \choose n}$ for positive integer $n$ is equal to the number of 2's in the prime factorization of $n$.

Your task is to write Python code to test this conjecture for as many positive integers as you can. See the self-assessment for more details.

Note: a `bitsoncount` function can be a one-liner in Python: `return bin(x).count('1')`



### Setup Functions

In [ ]:
# from math import gcd
import operator as op
from functools import reduce

# def nCk(n, k):
  # if k < 0 or k > n:
  #   return 0
  # else:
  #   result = 1
  #   d = 1
  #   g = 1
  #   m = min(k, n - k)
  #   while (d <= m):
  #     g = gcd(result, d)
  #     result = n * (result // g)
  #     result = (result // (d // g))
  #     n -= 1
  #     d += 1
  #   return result

'''
Davis suggested this function found here:
https://stackoverflow.com/questions/4941753/is-there-a-math-ncr-function-in-python
'''
def nCk(n, k):
  k = min(k, n-k)
  numer = reduce(op.mul, range(n, n-k, -1), 1)
  denom = reduce(op.mul, range(1, k+1), 1)
  return numer // denom

def count_twos(x):
  if x % 2 == 0 and x != 0:
    return 1 + count_twos(x // 2)
  else:
    return 0

bitsoncount = lambda x: bin(x).count('1')

In [ ]:
n = 26
print(count_twos(nCk(2*n, n)))
print(bitsoncount(n))

3
3


### Test a Bunch of Numbers

In [ ]:
def test_group(min, max):
  return [n for n in range(min, max) if count_twos(nCk(2*n, n)) != bitsoncount(n)]

def test_indiv(n):
  return [n] if count_twos(nCk(2*n, n)) != bitsoncount(n) else []

In [ ]:
import datetime
start_time = datetime.datetime.now()
print('Starting at ' + str(start_time))

# step = 2
n = 1
errors = []

try:
  while True:
    # errors += test_group(n, n + step)
    # n += step
    errors += test_indiv(n)
    n += 1
except:
  print(f'Verified up to {n}, finding {len(errors)} errors at {errors}. Time elapsed: {datetime.datetime.now() - start_time}')

Starting at 2021-02-05 08:31:28.406260
Verified up to 71672, finding 0 errors at []. Time elapsed: 1 day, 0:15:54.127286


In 24 hours, I was able to verify up to $n$ = 71,672. There were still no counterexamples found up to that point. I feel like having learned how to run code for long periods of time could be a useful skill in the future.

### Testing Different Step Values

In this portion, I ran the above code in 25 second intervals in order to determine which nCk function to use and whether or not to use list comprehensions for the above code.

Default nCk:
- Individual, n = 3152, t = 25.66721, n/t = 122.8
- step = 2, n = 3112, t = 25.444367, n/t = 122.3
- step = 10, n = 3130, t = 25.909244, n/t = 120.8
- step = 100, n = 3100, t = 25.449195, n/t = 121.8
- step = 1000, n = 3000, t = 25.415134, n/t = 118.0

Optimized nCk:
- Individual, n = 5274, t = 25.349994, n/t = 208.0
- step = 2, n = 5113, t = 25.52674, n/t = 200.3

## Basic Probability Theory Question
A dark room contains two barrels. The first barrel is filled with green marbles, the second is filled with a half-and-half mixture of green and blue marbles. So there's a 100% chance of choosing a green marble from the first barrel, and a 50% chance of choosing either color in the second barrel. You reach into one of the barrels (it's dark so you don't know which one) and select a marble at random. It's green. You select another. It's green too. You select a third, a fourth, a fifth, etc. Green each time. What is the *minimum* number of marbles you need to select to *exceed* a probability of 99% that you are picking them out of the all-green barrel? (Note that there are enough marbles so that the answer does not depend on how many marbles are in the second barrel.)


### Thoughts

Since these are the only two barrels, the following is true:

$$P(\text{All-Green}) = 1 - P(\text{Not-All-Green}) = 1 - P(\text{Half-and-Half})\text{.}$$

Since the second barrel has a 50% chance of choosing either color, the following is also true:

$$P(\text{Half-and-Half}) = 0.5^n\text{,}$$

where $n$ is the number of marbles pulled. If one wants $P(\text{All-Green}) > 0.99$, then one can do the following:

$$
1 - P(\text{Half-and-Half}) > 0.99\\
P(\text{Half-and-Half}) < 0.01\\
0.5^n < 0.01\\
n > log(0.01) / log(0.5)\\
n > 6.644\\
$$

Since the number of marbles ($n$) must be a positive integer, $n >= 7$.

## A Related But Deeper Basic Probability Theory Question
Take a deep breath. Suppose Shakespeare's account is accurate and Julius Caesar gasped "You too, Brutus" before breathing his last. What is the probability that you just inhaled a molecule that Julius Caesar exhaled in his dying breath?

Assume that after more than two thousand years the exhaled molecules are uniformly spread about the world and the vast majority are still free in the atmosphere. Assume further that there are $10^{44}$ molecules of air in the world, and that your inhaled quantity and Caesar's exhaled quantity were each about $2.2 \times 10^{22}$ molecules.
### Hint
If a number $x$ is small, then $(1 - x)$ is approximately equal to $e^{-x}$.


### Thoughts

The probability that a breath contains a molecule from Julius Caesar's last breath would be equal to 100% minus the probability of not sharing at least one molecule. In other words,

$$
P(\text{Breathing-Same}) = 1 - P(\text{Not-Breathing-Same})
$$

where $n$ is the number of molecules breathed in. This "not sharing" probability can be rephrased as the probability of not breathing in the first molecule, *and* not breathing in the second molecule, *and* not breathing in the third molecule, etc.

$$
P(\text{Not-Breathing-Same}) = P(\text{Not-Breathing-One-Molecule}) ^ n\text{,}
$$

where $n$ would be the number of molecules breathed in. Now, for any given molecule, the probability that it was *not* shared with Caesar's last breath is equal to 100% minus the probability that it was.

$$
P(\text{Not-Breathing-One-Molecule}) = 1 - P(\text{Breathing-One-Molecule})
$$

The probability that a given molecule was shared with Caesar's last breath can be found as follows:

$$
P(\text{Breathing-One-Molecule}) = n / w\text{,}
$$

where $n$ is the number of molecules exhaled and $w$ is the number of molecules of air in the world.

Thus we can combine these, then approximate using the identity $(1-x) \approx e^{-x}$:

$$
P(\text{Breathing-Same}) = 1 - ((1 - n / w)^n)\\
= 1 - ((1 - (2.2 \times 10^{22} / 10^{44})) ^ {2.2 \times 10^{22}})\\
\approx e^{-((1 - (2.2 \times 10^{22} / 10^{44})) ^ {2.2 \times 10^{22}})}\\
\approx e^{-((e^{-(2.2 \times 10^{22} / 10^{44})}) ^ {2.2 \times 10^{22}})}\\
\approx e^{-(e^{-(2.2 \times 10^{22} * 2.2 \times 10^{22} / 10^{44})})}\\
\approx e^{-(e^{-4.84})}
$$

In [ ]:
from math import e

e**(-(e**(-4.84)))

0.9921241244694242

There is an approximate chance of 99.2% that a given breath shares a molecule with Caesar's final breath.

## What is True?
Assess yourself on how you did using the checkboxes below. Check a box by putting an 'X' in it only if it is warranted.


### What is true of my experience in general?
(5 points each, 15 points total)
- [X] I had fun.
  - It felt similar to solving puzzles when I was simplifying equations and thinking through probabilities.
- [X] I learned something new.
  - I learned how to run a piece of code for a long period of time
- [X] I achieved something meaningful, or something I can build upon at a later time.
  - I got a refresher on probabilities.
  - I got some practice optimizing functions before running them for long periods of time.
  - I feel like I solved the puzzles given to me.

### What is true of my report on what I learned?
(5 points each, 25 points total)
- [X] I wrote a sufficient number of well-written sentences.
  - I feel my thought process has been laid out fairly well.
- [X] My report is free of "mechanical infelicities" (misspelled words, grammatical errors, punctuation errors, etc.).
  - I read through my smaller sections, then put my larger section through Grammarly.
- [X] I reported on any connections I found between this investigation and something I already know.
  - I had taken an Engineering Statistics course a few years ago, and many of the problems in this investigation reminded me of problems from that class.
  - The time I spent trying to optimize my code reminded me of the "parameter tuning" that I did in Machine Learning and Data Mining.
  - The approximation of $(1 - x) \approx e^{-x}$ for small values of x reminded me of using the Taylor Polynomials to approximate $sin(x)$ for small values of x.
- [X] I reported who were and what contribution each of my collaborators made.
  - See the page header and the comment block in my function declarations.
- [X] I reported on how many numbers I was able to verify with a time/computation budget of 24 hours (in a row).
  - See the section: "Test a Bunch of Numbers"


### What is true about my answers?
(15 points each, 60 points total)
- [X] I figured out how to run a Python program continuously for at least 24 hours.
  - The code output should still be there.
- [X] I refrained from printing out anything except the highest number I verified, knowing that printing just slows a program down.
  - The code output should still be there.
- [ ] I got the right answer for the first probability theory question.
  - I'm not sure how to verify this.
- [ ] I got the right answer for the second probability theory question.
  - I'm not sure how to verify this.
